kv_store/lww_kvs/kvs_ebs_server.cpp

```diff
@@ -1058,6 +1050,149 @@ int main(int argc, char* argv[]) {
+      cout << "Received replication factor change request\n";
+
+      string placement_req = zmq_util::recv_string(&replication_factor_change_puller);
+      communication::Placement_Request req;
+      req.ParseFromString(placement_req);
+
+      unordered_map<master_node_t, unordered_set<string>, node_hash> node_map;
+      // a set of keys to send to proxy to perform cross tier gossip
+      unordered_set<string> key_to_proxy;
+      // keep track of which key should be removed
+      unordered_map<string, bool> key_remove_map;
+
+      // for every key, update the replication factor and 
+      // check if the node is still responsible for the key
+      for (int i = 0; i < req.tuple_size(); i++) {
+        string key = req.tuple(i).key();
+
+        // first, check whether the node is responsible for the key before the replication factor change
+        auto result = responsible<master_node_t, crc32_hasher>(key, placement[key].global_ebs_replication_, global_ebs_hash_ring, mnode.id_);
+        // update placement info
+        placement[key].global_memory_replication_ = req.tuple(i).global_memory_replication();
+        placement[key].global_ebs_replication_ = req.tuple(i).global_ebs_replication();
+        // proceed only if the node is responsible for the key before the replication factor change
+        if (result.first) {
+          // check the global ring and request worker addresses from other node's master thread
+          auto pos = global_ebs_hash_ring.find(key);
+          for (int i = 0; i < placement[key].global_ebs_replication_; i++) {
+            if (pos->second.ip_.compare(ip) != 0) {
+              node_map[pos->second].insert(key);
+            }
+            if (++pos == global_ebs_hash_ring.end()) {
+              pos = global_ebs_hash_ring.begin();
+            }
+          }
+
+          // check if the key has replica on the memory tier
+          if (placement[key].global_memory_replication_ != 0) {
+            key_to_proxy.insert(key);
+          }
+
+          // check if the key has to be removed under the new replication factor
+          result = responsible<master_node_t, crc32_hasher>(key, placement[key].global_ebs_replication_, global_ebs_hash_ring, mnode.id_);
+          if (result.first) {
+            key_remove_map[key] = false;
+          } else {
+            key_remove_map[key] = true;
+          }
+        }
+      }
+
+      // initialize the redistribution map (key is the address of ebs worker thread)
+      unordered_map<string, redistribution_address*> redistribution_map;
+
+      for (auto map_iter = node_map.begin(); map_iter != node_map.end(); map_iter++) {
+        // get key address
+        communication::Key_Response resp = get_key_address<ebs_key_info>(map_iter->first.key_exchange_connect_addr_, "", map_iter->second, key_address_requesters, placement);
+        // for each key in the response
+        for (int i = 0; i < resp.tuple_size(); i++) {
```

Comment:
The code here is also duplicated in the `if key_to_proxy.size() != 0` block below, and I think in a couple other places.